### PR TITLE
CB-15466 Adjustments to support Azure FreeIPA HA upgrade

### PR DESCRIPTION
**Pull request by lacikaaa**
On Azure a 3 instance FreeIPA HA upgrade is taking more than 1 hour, which caused multiple timeouts.

Modifications in production code:
- Operation timeout increased in case of `UPGRADE` type of operation
- `operation` table's index was a mess, cleaned it up, and added only the necessary

Test code:
- Removed redundant class for `TimeoutChecker`
- Upgrade polling for FreeIPA is now polled for 2 hours, instead of attempt number
- AbsolutTimeoutChecker is available in test framework next to attempt based. It can be configured using `RunningParameter`

See detailed description in the commit message.

### CODE REVIEW

**Comment by lacikaaa** [comment]
retest this please

File: freeipa/src/main/resources/schema/app/20211214100402_CB-15466_cleanup_operation_indexes.sql [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+DROP INDEX IF EXISTS syncoperation_id_idx;
+
+DROP INDEX IF EXISTS syncoperation_operationid_idx;
+
+DROP INDEX IF EXISTS syncoperation_accountid_endtime_idx;
+
+CREATE INDEX IF NOT EXISTS operation_start_end_type_idx ON operation (starttime, endtime, operationtype)
+
+-- //@UNDO
+-- SQL to undo the change goes here.
```
Comment (topolyai5) [inline freeipa/src/main/resources/schema/app/20211214100402_CB-15466_cleanup_operation_indexes.sql:17]: Does not need the undo part?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/ResourceAwait.java [hunk 2]
```diff
@@ -37,7 +37,12 @@
                 client.waiterService().waitObject(new WaitFailedChecker<>(), waitObject, testContext, pollingInterval, maxRetry, maxRetryCount);
                 entity.refresh();
             } else {
-                client.waiterService().waitObject(new WaitOperationChecker<>(), waitObject, testContext, pollingInterval, maxRetry, maxRetryCount);
+                if (runningParameter != null && runningParameter.getTimeoutChecker() != null) {
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/ResourceAwait.java:40]: Could you please add this to the InstanceAwait?

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java [hunk 3]
```diff
@@ -6,14 +6,16 @@
...
+import com.sequenceiq.cloudbreak.polling.TimeoutChecker;
 import com.sequenceiq.it.cloudbreak.context.TestContext;
 
 public class WaitService<T extends WaitObject> {
     private static final Logger LOGGER = LoggerFactory.getLogger(WaitService.class);
 
     public Result<WaitResult, Exception> waitObject(StatusChecker<T> statusChecker, T t, TestContext testContext, Duration interval, int maxAttempts,
             int maxRetryCount) {
-        return waitObject(statusChecker, t, testContext, interval, new TimeoutChecker(maxAttempts), maxRetryCount);
+        return waitObject(statusChecker, t, testContext, interval, new AttemptBasedTimeoutChecker(maxAttempts), maxRetryCount);
```
Comment (topolyai5) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java:18]: Is it necessary to change the default time checker?

**Review by topolyai5 — DISMISSED** [review]

File: freeipa/src/main/resources/schema/app/20211214100402_CB-15466_cleanup_operation_indexes.sql (hunk 1 above)
Comment (lacikaaa) [reply freeipa/src/main/resources/schema/app/20211214100402_CB-15466_cleanup_operation_indexes.sql:17]: I'm adding the undo part for the new index, but not for cleaning the old ones, as they are not useful in any way or indexing the same column(s)

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 4]
```diff
@@ -87,7 +90,7 @@ public void testHAFreeIpaInstanceUpgrade(TestContext testContext) {
                 .given(FreeIpaOperationStatusTestDto.class)
                 .withOperationId(((FreeIpaTestDto) testContext.get(freeIpa)).getOperationId())
                 .then((tc, testDto, freeIpaClient) -> testFreeIpaAvailabilityDuringUpgrade(tc, testDto, freeIpaClient, freeIpa))
-                .await(COMPLETED)
+                .await(COMPLETED, waitForFlow().withWaitForFlow(Boolean.FALSE).withTimeoutChecker(new AbsolutTimeBasedTimeoutChecker(TWO_HOURS_IN_SEC)))
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java:93]: this shouldn't have to do anything with a flow, as it polls an operation, which in this case is a flowchain in FMS

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/ResourceAwait.java (hunk 2 above)
Comment (lacikaaa) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/ResourceAwait.java:40]: sure

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java (hunk 3 above)
Comment (lacikaaa) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/util/wait/service/WaitService.java:18]: it's not changed, it's renamed. `TimeoutChecker` is an interface now, etc

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/RunningParameter.java [hunk 5]
```diff
@@ -217,4 +224,9 @@ public static RunningParameter waitForFlowFail() {
         return new RunningParameter().withWaitForFlowFail();
     }
 
+    public RunningParameter withTimeoutChecker(TimeoutChecker timeoutChecker) {
```
[the remark this answers is not in the document]
Comment (lacikaaa) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/RunningParameter.java:227]: sure

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/operation/OperationTimeoutService.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+@Service
+public class OperationTimeoutService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(OperationTimeoutService.class);
+
+    @Value("${freeipa.operation.cleanup.timeout-millis}")
+    private  Long operationTimeout;
+
+    @Value("${freeipa.operation.cleanup.upgrade.timeout-millis}")
+    private  Long upgradeOperationTimeout;
```
Comment (horadla23) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/operation/OperationTimeoutService.java:25]: small nit: two spaces after `private`, not sure why checkstyle missed it

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/RunningParameter.java [hunk 7]
```diff
@@ -258,4 +265,13 @@ public static RunningParameter pollingInterval(Duration pollingInterval) {
     public static RunningParameter ignoredStatues(Set<Enum<?>> ignoredStatuses) {
         return new RunningParameter().withIgnoredStatues(ignoredStatuses);
     }
+
+    public RunningParameter withTimeoutChecker(TimeoutChecker timeoutChecker) {
+        this.timeoutChecker = timeoutChecker;
+        return this;
+    }
+
+    public static RunningParameter timoutChecker(TimeoutChecker timeoutChecker) {
```
Comment (horadla23) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/context/RunningParameter.java:274]: typo

**Review by horadla23 — APPROVED** [review]

**Pull request merged**